### PR TITLE
ejb-throws-exception quickstart

### PR DESCRIPTION
**Pull request by bmaxwell**

### CODE REVIEW

**Comment by pmuir** [comment]
@bmaxwell please ping back when you've fixed for Sande here (actually make a comment) and I'll do a review.

**Comment by bmaxwell** [comment]
@sgilda I made changes to the README.md, did they show up automatically on the pull request?

**Comment by sgilda** [comment]
Hi Brad, I'll try to look at this one today. :-)

**Comment by sgilda** [comment]
Hi Brad,
Your README file is perfect! Very informative. I wish they were all this detailed. I also tested this and it works great!

The only comments I have are related to the POM files. Since you wrote this, there have been a few changes.
- We have created BOM files and now use properties to define versions. Then we use the properties within  the dependencies in the file. This makes it easier for us to automate updates of version numbers as they change and we don't have to search through the text to find them.
- We fixed a bug where version numbers were appended to archive names. You can find the fixes for ejb-in-ear in this commit: 
      https://github.com/jboss-jdf/jboss-as-quickstart/commit/e87e865c52229f6ae8308f4265db14242b9849a2

You might want to rebase your branch with the upstream change to make it easier to see what has changed. 

Let me know if you need any help.
Sande

File: ejb-throws-exception/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+2. `web`: This project contains the JSF pages and the CDI managed bean.  The CDI Managed Bean (GreeterBean) will be bound to the JSF page (index.xhtml) and will invoke the GreeterEJB and display the response back from the EJB.  The GreeterBean catches the custom Exception (GreeterException) thrown by GreeterEJB and displays the Exception message in the response text on the JSF page.
+
+3. `ear`: This project builds the EAR artifact and pulls in the ejb, web, and client artifacts.
+
+4. `ejb-api`: This project builds the ejb-api library artifact which is used by the ejb, web, as well as remote client artifacts. The ejb-api directory contains the EJB interfaces, custom exceptions the EJB throws and any other transfer objects which the EJB may receive or send back to the client. The EJB interfaces, custom exceptions and other transfer objects are split into a separate jar which are packaged in the ear/lib which allows all sub deployments of the ear to see the classes of the ejb-api jar in their classpath.  This is also useful for remote clients.  The ejb-api jar can be distributed to a remote client and give the remote clients the classes that are needed to interact with the EJB
```
Comment (sgilda) [inline ejb-throws-exception/README.md]: This line is a little wordy: 
The EJB interfaces, custom exceptions and other transfer objects are split into a separate jar which are packaged in the ear/lib which allows all sub deployments of the ear to see the classes of the ejb-api jar in their classpath.  

 suggest:
The EJB interfaces, custom exceptions, and other transfer objects are split into a separate jar which is packaged in the ear/lib. This allows all sub deployments of the ear to see the classes of the ejb-api jar in their classpath.

**Comment by sgilda** [comment]
@bmaxwell :  Look at the kitchensink-ear or ejb-in-ear to see how to use peroperties, prevent version numbers from appearing in the archives, and for general structure.

**Comment by sgilda** [comment]
@bmaxwell : Your POM files still build the archives with the version number appended to the name, for example:

```
    jboss-as-ejb-throws-exception-ejb-7.1.2-SNAPSHOT
```

I will try to get it to work and then add comments to the files.

File: ejb-throws-exception/ear/pom.xml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                <artifactId>maven-ear-plugin</artifactId>
+                <version>${version.ear.plugin}</version>
+                <configuration>
+                    <!-- Tell Maven we are using Java EE 6 -->
+                    <version>${version.javaee}</version>
+                    <!-- Use Java EE ear libraries as needed. Java EE ear 
+                        libraries are in easy way to package any libraries needed in the ear, and 
+                        automatically have any modules (EJB-JARs and WARs) use them -->
+                    <defaultLibBundleDir>lib</defaultLibBundleDir>
+                </configuration>
```
Comment (sgilda) [inline ejb-throws-exception/ear/pom.xml:71]: Insert the following line after defaultLibBundleDir and before the ending configuration tag:

`<fileNameMapping>no-version</fileNameMapping>`

File: ejb-throws-exception/ear/pom.xml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            <!-- The JBoss AS plugin deploys your ear to a local JBoss AS 
+                container -->
+            <!-- Due to Maven's lack of intelligence with EARs we need to 
+                configure the jboss-as maven plugin to skip deployment for all modules. We 
+                then enable it specifically in the ear module. -->
+            <plugin>
+                <groupId>org.jboss.as.plugins</groupId>
+                <artifactId>jboss-as-maven-plugin</artifactId>
+                <configuration>
+                    <skip>false</skip>
```
Comment (sgilda) [inline ejb-throws-exception/ear/pom.xml:83]: Insert the following line after the configuration tag and before the skip false tag:

`<filename>${project.artifactId}.ear</filename>`

`

File: ejb-throws-exception/ejb-api/pom.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.jboss.spec.javax.ejb</groupId>
+            <artifactId>jboss-ejb-api_3.1_spec</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+</project>
```
Comment (sgilda) [inline ejb-throws-exception/ejb-api/pom.xml:55]: Insert the following before the end project tag (the indentation doesn't appear correctly unless you go into edit mode on the comment):

`<build>
        <finalName>${project.artifactId}</finalName>
    </build>
`

File: ejb-throws-exception/pom.xml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            </dependency>
+
+            <!-- Define the version of the WAR so that we don't need to repeat 
+                ourselves in every module -->
+            <dependency>
+                <groupId>org.jboss.as.quickstarts</groupId>
+                <artifactId>jboss-as-ejb-throws-exception-web</artifactId>
+                <version>${project.version}</version>
+                <type>war</type>
+            </dependency>
```
Comment (sgilda) [inline ejb-throws-exception/pom.xml:109]: I'm not sure if this is necessary, but the other EAR quickstarts add `<scope>compile</scope` after the `<type>war</type>` element.

File: ejb-throws-exception/web/pom.xml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        <artifactId>jboss-as-ejb-throws-exception</artifactId>
+        <version>7.1.2-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>jboss-as-ejb-throws-exception-web</artifactId>
+    <packaging>war</packaging>
+
+    <name>JBoss AS Quickstarts: EJB throws exception (WAR)</name>
+
+    <dependencies>
```
Comment (sgilda) [inline ejb-throws-exception/web/pom.xml:37]: Insert the licence before dependencies"

`
    <licenses>
        <license>
            <name>Apache License, Version 2.0</name>
            <distribution>repo</distribution>
            <url>http://www.apache.org/licenses/LICENSE-2.0.html</url>
        </license>
    </licenses>
`

File: ejb-throws-exception/web/pom.xml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        <dependency>
+            <groupId>org.jboss.spec.javax.ejb</groupId>
+            <artifactId>jboss-ejb-api_3.1_spec</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
```
Comment (sgilda) [inline ejb-throws-exception/web/pom.xml:82]: Insert the finalName before the plugins:

`<finalName>${project.artifactId}</finalName>`

File: ejb-throws-exception/ejb/pom.xml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+        <dependency>
+            <groupId>org.jboss.as.quickstarts</groupId>
+            <artifactId>jboss-as-ejb-throws-exception-ejb-api</artifactId>
+        </dependency>
+        
+    </dependencies>
+
+    <build>
+        <plugins>
```
Comment (sgilda) [inline ejb-throws-exception/ejb/pom.xml:85]: Insert the finalName before the plugins:

`<finalName>${project.artifactId}</finalName>`

**Comment by sgilda** [comment]
@bmaxwell : I think I marked all the changes required to get rid of version numbers in the archive names. Let me know if you have any questions.

**Comment by bmaxwell** [comment]
Pushed again.  I build and the version numbers are in the build ear,jars or war.

**Comment by sgilda** [comment]
@bmaxwell : Did you mean the version numbers are no longer appended to the archive names? This looks good to me. They now build with the names ejb-throws-exception-ear.ear, jboss-as-ejb-throws-exception-ejb.jar, jboss-as-ejb-throws-exception-ejb-api.jar , and jboss-as-ejb-throws-exception-web.war.

@pmuir: Would you like to review?

**Comment by bmaxwell** [comment]
@sgilda yes sorry, busy week, apparently I'm losing it :)  Yes, I made the changes and the version number are not in the ear,war,jars.

**Comment by sgilda** [comment]
@bmaxwell : I'm glad to know I'm not the only one losing it.
After @pmuir  reviews this one, I will merge it.

File: ejb-throws-exception/ejb-api/pom.xml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        <dependency>
+            <groupId>org.jboss.spec.javax.ejb</groupId>
+            <artifactId>jboss-ejb-api_3.1_spec</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <finalName>${project.artifactId}</finalName>
+    </build>
```
Comment (sgilda) [inline ejb-throws-exception/ejb-api/pom.xml:53]: I think you might also need to add the following:
`
     <plugins>
         <plugin>
            <artifactId>maven-jar-plugin</artifactId>
            <version>${version.jar.plugin}</version>
         </plugin>
      </plugins>
`
See the helloworld-jms quickstart for an example.

File: ejb-throws-exception/ejb-api/pom.xml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        <groupId>org.jboss.as.quickstarts</groupId>
+        <artifactId>jboss-as-ejb-throws-exception</artifactId>
+        <version>7.1.2-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>jboss-as-ejb-throws-exception-ejb-api</artifactId>
+    <packaging>jar</packaging>
+
+    <name>ejb-throws-exception-ejb-api</name>
+
```
Comment (sgilda) [inline ejb-throws-exception/ejb-api/pom.xml:25]: I think you might need the following also:
`
    <licenses>
        <license>
            <name>Apache License, Version 2.0</name>
            <distribution>repo</distribution>
            <url>http://www.apache.org/licenses/LICENSE-2.0.html</url>
        </license>
    </licenses>
`
`
    <properties>
        <version.jar.plugin>2.2</version.jar.plugin>
    </properties>
`

**Comment by sgilda** [comment]
@pmuir : Would you like to review this one last time?

**Comment by pmuir** [comment]
Yes, next week :-)

**Comment by pmuir** [comment]
Looks good. I've merged it. Thanks @bmaxwell!

**Pull request closed without merging**